src/tests/unit/tests/background/tab-context-manager.test.ts

```diff
@@ -0,0 +1,121 @@
+import { Interpreter } from 'background/interpreter';
+import { TabContext } from 'background/tab-context';
+import { TabContextFactory } from 'background/tab-context-factory';
+import { TabContextManager } from 'background/tab-context-manager';
+import { Message } from 'common/message';
+import { IMock, It, Mock, Times } from 'typemoq';
+import { DictionaryNumberTo } from 'types/common-types';
+import { BrowserAdapter } from '../../../../common/browser-adapters/browser-adapter';
+
+describe(TabContextManager, () => {
+    const tabId = 4;
+    const persistStoreData = false;
+
+    let mockBroadcasterFactoryMock: IMock<BrowserMessageBroadcasterFactory>;
+    let mockTabContextFactory: IMock<TabContextFactory>;
+    let mockBrowserAdapter: IMock<BrowserAdapter>;
+    let mockDetailsViewController: IMock<ExtensionDetailsViewController>;
+    let tabToContextMap: DictionaryNumberTo<TabContext>;
+    let tabContextMock: IMock<TabContext>;
+    let interpreterMock: IMock<Interpreter>;
+
+    let testSubject: TabContextManager;
+
+    beforeEach(() => {
+        mockBroadcasterFactoryMock = Mock.ofType<BrowserMessageBroadcasterFactory>();
+        mockBrowserAdapter = Mock.ofType<BrowserAdapter>();
+        mockDetailsViewController = Mock.ofType<ExtensionDetailsViewController>();
+        tabToContextMap = {};
+        mockTabContextFactory = Mock.ofType<TabContextFactory>();
+        tabContextMock = Mock.ofType<TabContext>();
+        interpreterMock = Mock.ofType<Interpreter>();
+
+        testSubject = new TabContextManager(
+            tabToContextMap,
+            mockBroadcasterFactoryMock.object,
+            mockBrowserAdapter.object,
+            mockDetailsViewController.object,
+            mockTabContextFactory.object,
+            persistStoreData,
+        );
+    });
+
+    afterEach(() => {
+        mockBroadcasterFactoryMock.verifyAll();
+        mockTabContextFactory.verifyAll();
+        tabContextMock.verifyAll();
+        interpreterMock.verifyAll();
+    });
+
+    it('Creates new tab context', () => {
+        const broadcasterMock = Mock.ofType<MessageBroadcaster>();
+        mockBroadcasterFactoryMock
+            .setup(m => m.createTabSpecificBroadcaster(tabId))
+            .returns(() => broadcasterMock.object)
+            .verifiable();
+        mockTabContextFactory
+            .setup(m =>
+                m.createTabContext(
+                    broadcasterMock.object,
+                    mockBrowserAdapter.object,
+                    mockDetailsViewController.object,
+                    tabId,
+                    persistStoreData,
+                ),
+            )
+            .returns(() => tabContextMock.object)
+            .verifiable();
+
+        testSubject.addTabContextIfNotExists(tabId);
+
+        expect(tabToContextMap[tabId]).toBe(tabContextMock.object);
+    });
+
+    it('Does not recreate tab context if already exists', () => {
```

Comment:
Again, not sure if we ever get into this state, but might be worthwhile adding a test case for when `tabToContextMap[tabId] = undefined`, if nothing else to define the behavior.